timeseries/src/autogluon/timeseries/models/autogluon_tabular/tabular_model.py

```diff
@@ -106,25 +109,62 @@ def _get_features_dataframe(
             If provided, features will be generated only for the last `last_k_values` timesteps of each time series.
         """
 
-        def get_lag_features_and_target(group):
-            timestamp = group.index.get_level_values(TIMESTAMP)
-            lag_columns = {f"lag_{idx}": group.shift(idx).values.ravel() for idx in self._lag_indices}
-            features = pd.DataFrame(lag_columns, index=timestamp)
+        def get_lags(df: pd.DataFrame, lag_indices: List[int]) -> pd.DataFrame:
+            shifted = [df.shift(idx).add_suffix(f"_lag_{idx}") for idx in lag_indices]
+            return pd.concat(shifted, axis=1)
+
+        def apply_mask(
+            df: pd.DataFrame, num_hidden: np.ndarray, lag_indices: np.ndarray, num_columns: int = 1
+        ) -> pd.DataFrame:
+            if num_columns > 1:
+                lag_indices = np.repeat(lag_indices, num_columns)
+            mask = num_hidden[:, None] >= lag_indices[None]  # shape [len(num_hidden), len(lag_indices) * num_columns]
+            df[mask] = np.nan
+            return df
+
+        def get_lag_features_and_target(time_series):
             # Starting from the end of the time series, mask the values as if the last `prediction_length` steps weren't observed
```

Comment:
is this "docstring" outdated? can we add docstrings to all of the nested functions? I suspect this method got a little too complicated and could use a refactor.